File: Algorithms/src/1081. Smallest Subsequence of Distinct Characters/Code/solution.cpp
class Solution
{
public:
    string smallestSubsequence(string s)
    {
        // Store how many times each character still appears
        vector<int> freq(26, 0);
        for (char c : s)
            freq[c - 'a']++;

        // Track whether a character is already inside the stack
        vector<bool> inStack(26, false);

        // This string works as our stack
        string st;

        for (char c : s)
        {
            // One occurrence of this character has now been processed
            freq[c - 'a']--;

            // Skip duplicate characters already included
            if (inStack[c - 'a'])
                continue;

            // Remove larger characters if they appear again later
            while (!st.empty() &&
                   st.back() > c &&
                   freq[st.back() - 'a'] > 0)
            {

                // Mark removed character as no longer inside the stack
                inStack[st.back() - 'a'] = false;
                st.pop_back();
            }

            // Add current character to the stack
            st.push_back(c);
            inStack[c - 'a'] = true;
        }

        // The stack already contains the final answer
        return st;
    }
};

File: Algorithms/src/1081. Smallest Subsequence of Distinct Characters/Code/solution.js
/**
 * @param {string} s
 * @return {string}
 */
var smallestSubsequence = function(s) {

    // Count remaining occurrences
    const freq = new Array(26).fill(0);
    for (const ch of s) {
        freq[ch.charCodeAt(0) - 97]++;
    }

    // Track characters already inside the stack
    const inStack = new Array(26).fill(false);

    // Array works as a stack
    const stack = [];

    for (const ch of s) {

        // Current occurrence has been processed
        freq[ch.charCodeAt(0) - 97]--;

        // Skip duplicates
        if (inStack[ch.charCodeAt(0) - 97]) {
            continue;
        }

        // Remove larger characters if they appear again later
        while (
            stack.length > 0 &&
            stack[stack.length - 1] > ch &&
            freq[stack[stack.length - 1].charCodeAt(0) - 97] > 0
        ) {
            inStack[stack.pop().charCodeAt(0) - 97] = false;
        }

        // Push current character
        stack.push(ch);
        inStack[ch.charCodeAt(0) - 97] = true;
    }

    // Join stack into the answer
    return stack.join("");
};